connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java

```diff
@@ -0,0 +1,258 @@
+                    break;
+                }
+            }
+        }
+        // add last step if last char is a dot
+        if (!path.isEmpty() && path.charAt(path.length() - 1) == DOT)
+            steps.add("");
+        return Collections.unmodifiableList(steps);
+    }
+
+    private static void failWhenIncompleteBacktickPair(String path, int backtickAt) {
+        throw new ConfigException("Incomplete backtick pair in path: [" + path + "],"
+                + " consider adding a backslash before backtick at position " + backtickAt
+                + " to escape it");
+    }
+
+    public String originalPath() {
+        return this.originalPath;
+    }
+
+
+    /**
+     * Access a {@code Field} at the current path within a schema {@code Schema}
+     * If field is not found, then {@code null} is returned.
+     */
+    public Field fieldFrom(Schema schema) {
+        if (schema == null) return null;
+
+        Schema current = schema;
+        for (String pathSegment : stepsWithoutLast()) {
+            final Field field = current.field(pathSegment);
+            if (field != null) {
+                current = field.schema();
+            } else {
+                return null;
+            }
+        }
+        return current.field(lastStep());
+    }
+
+    /**
+     * Access a value at the current path within a schema-based {@code Struct}
+     * If object is not found, then {@code null} is returned.
+     */
+    public Object valueFrom(Struct struct) {
+        if (struct == null) return null;
+
+        Struct current = struct;
+        for (String pathSegment : stepsWithoutLast()) {
+            // Check to see if the field actually exists
+            if (current.schema().field(pathSegment) == null) {
+                return null;
+            }
+            current = current.getStruct(pathSegment);
+        }
+
+        if (current.schema().field(lastStep()) != null) {
+            return current.get(lastStep());
+        } else {
+            return null;
+        }
+    }
+
+    List<String> stepsWithoutLast() {
+        return steps.subList(0, lastStepIndex());
+    }
+
+    /**
+     * Access a value at the current path within a schemaless {@code Map<String, Object>}.
+     * If object is not found, then {@code null} is returned.
+     */
+    @SuppressWarnings("unchecked")
+    public Object valueFrom(Map<String, Object> map) {
+        if (map == null) return null;
+
+        if (steps.size() == 1) {
+            return map.get(steps.get(0));
+        } else {
+            Map<String, Object> current = map;
+            for (int i = 0; i < steps.size(); i++) {
+                if (current == null) {
+                    return null;
+                }
+                if (i == lastStepIndex()) {
+                    return current.get(steps.get(i));
+                } else {
+                    current = (Map<String, Object>) current.get(steps.get(i));
+                }
+            }
+        }
+        return null;
```

Comment:
We can clean this up a bit:
```suggestion
        if (map == null) return null;

        Map<String, Object> current = map;
        for (String step : stepsWithoutLast()) {
            current = requireMapOrNull(current.get(step), "nested field access");

            if (current == null)
                return null;
        }
        return current.get(lastStep());
```

Using `requireMapOrNull` (statically imported from `org.apache.kafka.connect.transforms.util.Requirements`) makes this more type-safe, and leveraging `stepsWithoutLast` and `lastStep` helps simplify the logic.

Reply:
Oh, and the `@SuppressWarnings` annotation can be removed if this works.